lib/redis/ick.rb

```diff
@@ -301,16 +273,179 @@ def ickcommit(ick_key,*members)
+    # committed regardless of how low scores in the pset might be.
+    #
+    # @return a list of up to reserve_size pairs, similar to
+    # Redis.current.zrange() withscores: [ message, score ]
+    # representing the lowest-scored elements from the producer set
+    # after the commit and reserve operations.
+    #
+    def ickexchange(ick_key,reserve_size,*commit_members,backwash: false)
+      if !ick_key.is_a?(String)
+        raise ArgumentError, "bogus non-String ick_key #{ick_key}"
+      end
+      if !reserve_size.is_a?(Integer)
+        raise ArgumentError, "bogus non-Integer reserve_size #{reserve_size}"
+      end
+      if reserve_size < 0
+        raise ArgumentError, "bogus negative reserve_size #{reserve_size}"
+      end
+      _statsd_increment('profile.ick.ickexchange.calls')
+      _statsd_timing('profile.ick.ickexchange.reserve_size',reserve_size)
+      _statsd_timing(
+        'profile.ick.ickexchange.commit_members',
+        commit_members.size
+      )
+      raw_results = nil
+      _statsd_time('profile.ick.time.ickexchange') do
+        raw_results = _eval(
+          LUA_ICKEXCHANGE,
+          ick_key,
+          reserve_size,
+          backwash ? 'backwash' : false,
+          commit_members
+        )
+      end
+      _postprocess(raw_results,Skip0ThenFloatifyPairs)
+    end
+
+    # Postprocessing done on the LUA_ICKEXCHANGE results for both
+    # ickreserve and ickexchange.
+    #
+    # results are num_committed followed by N message-and-score
+    # pairs.
+    #
+    # We do results[1..-1] to skip the first element, num_committed.
+    #
+    # On the rest, we floatify the scores to convert from Redis
+    # number-as-string limitation to Ruby Floats.
+    #
+    # This is similar to to Redis::FloatifyPairs:
+    #
+    # https://github.com/redis/redis-rb/blob/master/lib/redis.rb#L2887-L2896
+    #
+    Skip0ThenFloatifyPairs = lambda do |results|
+      results[1..-1].each_slice(2).map do |m_and_s|
+        [ m_and_s[0], ::Redis::Ick._floatify(m_and_s[1]) ]
+      end
+    end
+
+    # Calls back to block with the results.
+    # 
+    # If raw_results is a Redis::Future, callback will be deferred
+    # until the future is expanded.
+    #
+    # Otherwise, callback will happen immediately.
+    #
+    def _postprocess(raw_results,callback)
+      if raw_results.is_a?(Redis::Future)
+        #
+        # Redis::Future have a built-in mechanism for calling a
+        # transformation on the raw results.
+        #
+        # Here, we monkey-patch not the Redis::Future class, but just
+        # this one raw_results object.  We give ourselves a door to
+        # set the post-processing transformation.
+        #
+        # The ransformation will be called only once when the real
+        # results are materialized.
+        #
+        class << raw_results
+          def transformation=(transformation)
+            raise "transformation collision" if @transformation
```

Comment:
Good safety check, I wonder if you could preserve the original transformation and chain it with the new one somehow

Reply:
I'm pretty sure that none of the Redis methods which are called via `_eval` result in transformations.  Those are present for things like `Redis.zscore` for which it needs to call `Redis::Floatify`, but no blocks are provided for any of `Redis.eval`, `Redis.evalsha`, etc.